File: src/Sems.Api/Modules/Payments/Interfaces/PaymentResources.cs
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;
using Sems.Api.Modules.Payments.Domain.Model;

namespace Sems.Api.Modules.Payments.Interfaces;

/// <summary>Contrato JSON del modulo de pagos, en snake_case como el original.</summary>
public static class PaymentResources
{
    // ------------------------------------------------------------- peticiones

    public sealed record ProcessPaymentRequest(
        [property: JsonPropertyName("subscription_id")]
        [property: Required(ErrorMessage = "is required")] string SubscriptionId,
        [property: JsonPropertyName("user_id")]
        [property: Required(ErrorMessage = "is required")] string UserId,
        [property: JsonPropertyName("payment_method_id")]
        [property: Required(ErrorMessage = "is required")] string PaymentMethodId,
        [property: JsonPropertyName("amount")] double Amount,
        [property: JsonPropertyName("currency")] string? Currency,
        [property: JsonPropertyName("payment_method")] string? PaymentMethod);

    public sealed record RegisterPaymentMethodRequest(
        [property: JsonPropertyName("user_id")]
        [property: Required(ErrorMessage = "is required")] string UserId,
        [property: JsonPropertyName("type")] string? Type,
        [property: JsonPropertyName("stripe_payment_method_id")]
        [property: Required(ErrorMessage = "is required")] string StripePaymentMethodId,
        [property: JsonPropertyName("is_default")] bool IsDefault);

    /// <summary>Cuerpo que envia la aplicacion web para abrir Stripe Checkout.</summary>
    public sealed record CreateCheckoutRequest(
        [property: JsonPropertyName("user_id")]
        [property: Required(ErrorMessage = "is required")] string UserId,
        [property: JsonPropertyName("subscription_id")] string? SubscriptionId,
        [property: JsonPropertyName("plan_name")] string? PlanName,
        [property: JsonPropertyName("amount")] double Amount,
        [property: JsonPropertyName("currency")] string? Currency,
        [property: JsonPropertyName("success_url")]
        [property: Required(ErrorMessage = "is required")] string SuccessUrl,
        [property: JsonPropertyName("cancel_url")]
        [property: Required(ErrorMessage = "is required")] string CancelUrl);

    // -------------------------------------------------------------- respuestas

    public sealed record PaymentResponse(
        [property: JsonPropertyName("payment_id")] string PaymentId,
        [property: JsonPropertyName("subscription_id")] string? SubscriptionId,
        [property: JsonPropertyName("user_id")] string UserId,
        [property: JsonPropertyName("payment_method_id")] string? PaymentMethodId,
        [property: JsonPropertyName("amount")] double Amount,
        [property: JsonPropertyName("currency")] string Currency,
        [property: JsonPropertyName("status")] PaymentStatus Status,
        [property: JsonPropertyName("payment_method")] string? PaymentMethod,
        [property: JsonPropertyName("stripe_payment_intent_id")] string? StripePaymentIntentId,
        [property: JsonPropertyName("paid_at")] DateTime? PaidAt,
        [property: JsonPropertyName("created_at")] DateTime CreatedAt)
    {
        public static PaymentResponse From(Payment p) => new(p.PaymentId.ToString(),
            p.SubscriptionId?.ToString(), p.UserId.ToString(), p.PaymentMethodId?.ToString(),
            p.Amount, p.Currency, p.Status, p.PaymentMethod, p.StripePaymentIntentId, p.PaidAt,
            p.CreatedAt);
    }

    public sealed record InvoiceResponse(
        [property: JsonPropertyName("invoice_id")] string InvoiceId,
        [property: JsonPropertyName("payment_id")] string PaymentId,
        [property: JsonPropertyName("invoice_number")] string InvoiceNumber,
        [property: JsonPropertyName("issued_at")] DateTime IssuedAt,
        [property: JsonPropertyName("total_amount")] double TotalAmount,
        [property: JsonPropertyName("pdf_url")] string? PdfUrl)
    {
        public static InvoiceResponse From(Invoice i) => new(i.InvoiceId.ToString(),
            i.PaymentId.ToString(), i.InvoiceNumber, i.IssuedAt, i.TotalAmount, i.PdfUrl);
    }

    public sealed record ProcessPaymentResponse(
        [property: JsonPropertyName("payment")] PaymentResponse Payment,
        [property: JsonPropertyName("invoice")] InvoiceResponse? Invoice);

    public sealed record PaymentMethodResponse(
        [property: JsonPropertyName("payment_method_id")] string PaymentMethodId,
        [property: JsonPropertyName("user_id")] string UserId,
        [property: JsonPropertyName("type")] string? Type,
        [property: JsonPropertyName("brand")] string? Brand,
        [property: JsonPropertyName("last4")] string? Last4,
        [property: JsonPropertyName("exp_month")] int ExpMonth,
        [property: JsonPropertyName("exp_year")] int ExpYear,
        [property: JsonPropertyName("stripe_payment_method_id")] string? StripePaymentMethodId,
        [property: JsonPropertyName("is_default")] bool IsDefault,
        [property: JsonPropertyName("created_at")] DateTime CreatedAt)
    {
        public static PaymentMethodResponse From(PaymentMethodEntity m) =>
            new(m.PaymentMethodId.ToString(), m.UserId.ToString(), m.Type, m.Brand, m.Last4,
                m.ExpMonth, m.ExpYear, m.StripePaymentMethodId, m.IsDefault, m.CreatedAt);
    }

    public sealed record CheckoutSessionResponse(
        [property: JsonPropertyName("session_id")] string? SessionId,
        [property: JsonPropertyName("url")] string? Url);
}

File: src/main/java/com/sems/payments/interfaces/rest/resources/PaymentResources.java
package com.sems.payments.interfaces.rest.resources;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import com.sems.payments.domain.model.entities.*;
import com.sems.payments.domain.model.valueobjects.PaymentStatus;
import jakarta.validation.constraints.NotBlank;
import java.time.Instant;

/**
 * Contrato JSON del modulo de pagos, en snake_case como el servicio original.
 */
public final class PaymentResources {

    private PaymentResources() {
    }

    // ------------------------------------------------------------- peticiones

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record ProcessPaymentRequest(
            @NotBlank(message = "is required") String subscriptionId,
            @NotBlank(message = "is required") String userId,
            @NotBlank(message = "is required") String paymentMethodId,
            double amount,
            String currency,
            String paymentMethod) {
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record RegisterPaymentMethodRequest(
            @NotBlank(message = "is required") String userId,
            String type,
            @NotBlank(message = "is required") String stripePaymentMethodId,
            boolean isDefault) {
    }

    /** Cuerpo que envia la aplicacion web para abrir Stripe Checkout. */
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record CreateCheckoutRequest(
            @NotBlank(message = "is required") String userId,
            String subscriptionId,
            String planName,
            double amount,
            String currency,
            @NotBlank(message = "is required") String successUrl,
            @NotBlank(message = "is required") String cancelUrl) {
    }

    // -------------------------------------------------------------- respuestas

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record PaymentResponse(
            String paymentId, String subscriptionId, String userId, String paymentMethodId,
            double amount, String currency, PaymentStatus status, String paymentMethod,
            String stripePaymentIntentId, Instant paidAt, Instant createdAt) {

        public static PaymentResponse from(Payment p) {
            return new PaymentResponse(p.getPaymentId().toString(),
                    p.getSubscriptionId() == null ? null : p.getSubscriptionId().toString(),
                    p.getUserId().toString(),
                    p.getPaymentMethodId() == null ? null : p.getPaymentMethodId().toString(),
                    p.getAmount(), p.getCurrency(), p.getStatus(), p.getPaymentMethod(),
                    p.getStripePaymentIntentId(), p.getPaidAt(), p.getCreatedAt());
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record InvoiceResponse(String invoiceId, String paymentId, String invoiceNumber,
                                  Instant issuedAt, double totalAmount, String pdfUrl) {

        public static InvoiceResponse from(Invoice i) {
            return new InvoiceResponse(i.getInvoiceId().toString(), i.getPaymentId().toString(),
                    i.getInvoiceNumber(), i.getIssuedAt(), i.getTotalAmount(), i.getPdfUrl());
        }
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record ProcessPaymentResponse(PaymentResponse payment, InvoiceResponse invoice) {
    }

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record PaymentMethodResponse(
            String paymentMethodId, String userId, String type, String brand, String last4,
            int expMonth, int expYear, String stripePaymentMethodId, boolean isDefault,
            Instant createdAt) {

        public static PaymentMethodResponse from(PaymentMethod m) {
            return new PaymentMethodResponse(m.getPaymentMethodId().toString(),
                    m.getUserId().toString(), m.getType(), m.getBrand(), m.getLast4(),
                    m.getExpMonth(), m.getExpYear(), m.getStripePaymentMethodId(),
                    m.isDefaultMethod(), m.getCreatedAt());
        }
    }

    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
    public record CheckoutSessionResponse(String sessionId, String url) {
    }
}
